linkerd/router/src/cache.rs

```diff
@@ -0,0 +1,89 @@
+use crate::NewService;
+use ahash::AHashMap;
+use parking_lot::Mutex;
+use std::{fmt::Debug, hash::Hash, marker::PhantomData, sync::Arc};
+
+/// A [`NewService`] that lazy builds stacks for each `K`-typed key.
+#[derive(Debug)]
+pub struct NewCache<K, N> {
+    inner: N,
+    _marker: PhantomData<fn(K)>,
+}
+
+/// A [`NewService`] that lazily builds an inner `S`-typed service for each
+/// `K`-typed target.
+#[derive(Debug)]
+pub struct Cache<K, N, S> {
+    new_inner: N,
+    services: Arc<Mutex<AHashMap<K, S>>>,
+}
+
+// === impl NewCache ===
+
+impl<K, N> NewCache<K, N> {
+    pub fn new(inner: N) -> Self {
+        Self {
+            inner,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<T, K, N, M> NewService<T> for NewCache<K, N>
+where
+    T: Clone,
+    N: NewService<T, Service = M>,
+    M: NewService<K>,
+{
+    type Service = Cache<K, M, M::Service>;
+
+    fn new_service(&self, target: T) -> Self::Service {
+        Cache::new(self.inner.new_service(target))
+    }
+}
+
+impl<K, N: Clone> Clone for NewCache<K, N> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            _marker: self._marker,
+        }
+    }
+}
+
+// === impl Cache ===
+
+impl<K, N, S> Cache<K, N, S> {
+    pub(super) fn new(new_inner: N) -> Self {
+        Self {
+            new_inner,
+            services: Default::default(),
+        }
+    }
+}
+
+impl<K, N> NewService<K> for Cache<K, N, N::Service>
+where
+    K: Eq + Hash + Clone,
+    N: NewService<K>,
+    N::Service: Clone,
+{
+    type Service = N::Service;
+
+    fn new_service(&self, key: K) -> Self::Service {
+        self.services
+            .lock()
+            .entry(key.clone())
```

Comment:
could be worth noting in the docs that `Cache`'s `NewService` impl will always clone the target type, even if there's already a cached service for it...